tests/transformations/state_fusion_test.py

```diff
@@ -3,17 +3,220 @@
+
+    assert sdfg.apply_transformations_repeated(StateFusion, strict=True) == 1
+
+
+def test_two_cc_fusion_together():
+    """ Two states, both with two connected components, fused to one CC. """
+    sdfg = dace.SDFG('state_fusion_test')
+    sdfg.add_array('A', [1], dace.int32)
+    sdfg.add_array('B', [1], dace.int32)
+    sdfg.add_array('C', [1], dace.int32)
+    state1, state2 = tuple(sdfg.add_state() for _ in range(2))
+    sdfg.add_edge(state1, state2, dace.InterstateEdge())
+
+    # First state
+    state1.add_edge(state1.add_tasklet('one', {}, {'a'}, 'a = 1'), 'a',
+                    state1.add_write('A'), None, dace.Memlet('A'))
+
+    t2 = state1.add_tasklet('two', {}, {'b', 'c'}, 'b = 2; c = 3')
+    state1.add_edge(t2, 'b', state1.add_write('B'), None, dace.Memlet('B'))
+    state1.add_edge(t2, 'c', state1.add_write('C'), None, dace.Memlet('C'))
+
+    # Second state
+    state2.add_edge(state2.add_read('B'), None,
+                    state2.add_tasklet('one', {'a'}, {}, ''), 'a',
+                    dace.Memlet('B'))
+
+    t2 = state2.add_tasklet('two', {'b', 'c'}, {'d', 'e'}, 'd = b + c; e = b')
+    state2.add_edge(state2.add_read('A'), None, t2, 'b', dace.Memlet('A'))
+    state2.add_edge(state2.add_read('C'), None, t2, 'c', dace.Memlet('C'))
+    state2.add_edge(t2, 'd', state2.add_write('A'), None, dace.Memlet('A'))
+    state2.add_edge(t2, 'e', state2.add_write('C'), None, dace.Memlet('C'))
+
+    assert sdfg.apply_transformations_repeated(StateFusion, strict=True) == 1
+
+
+# Data race avoidance tests
+def test_write_write_path():
+    """
+    Two states where both write to the same range of an array, but there is
+    a path between the write and the second write.
+    """
+    @dace.program
+    def state_fusion_test(A: dace.int32[20, 20]):
+        A += 1
+        tmp = A + 2
+        A[:] = tmp + 3
+
+    sdfg = state_fusion_test.to_sdfg(strict=False)
+    sdfg.apply_transformations_repeated(StateFusion, strict=True)
+    assert len(sdfg.nodes()) == 1
+
+
+def test_write_write_no_overlap():
+    """
+    Two states where both write to different ranges of an array.
+    """
+    N = dace.symbol('N', positive=True)
+
+    @dace.program
+    def state_fusion_test(A: dace.int32[N, N]):
+        A[0:N - 1, :] = 1
+        A[N - 1, :] = 2
+
+    sdfg = state_fusion_test.to_sdfg(strict=False)
+    sdfg.apply_transformations_repeated(StateFusion, strict=True)
+    assert len(sdfg.nodes()) == 1
+
+
+def test_read_write_no_overlap():
+    """
+    Two states where two separate CCs write and read to/from an array, but
+    in different ranges.
+    """
+    N = dace.symbol('N')
+
+    @dace.program
+    def state_fusion_test(A: dace.int32[N, N], B: dace.int32[N, N]):
+        A[:, 5:N] = 1
+        B[:, 3:6] = A[:, 0:3]
+
+    sdfg = state_fusion_test.to_sdfg(strict=False)
+    sdfg.apply_transformations_repeated(StateFusion, strict=True)
+    assert len(sdfg.nodes()) == 1
+
+
 if __name__ == '__main__':
```

Comment:
Do we have a test where both states do a write/read in an intermediate array (first write, then read, so that the array appears neither in the state inputs nor the outputs)? Furthermore, the states should not have any other dependency. This means that, if they were fused, the result would be 2 CCs, violating the data dependency on the intermediate array.

Reply:
Thanks, added tests to reflect.